backend/fms_core/viewsets/_fetch_data.py

```diff
@@ -333,6 +333,78 @@ def fetch_export_data(self, ids: List[int] =[]) -> Tuple[List, int]:
+
+        Args:
+            self: base_class(viewset) + fetch_data
+            ids: List of ids to select specific samples. Defaults to [].
+
+        Returns:
+            Returns a list of serialized sample metadata dictionary and a tuple of the names of the metadata for rendering purposes
+        """
+
+        super().fetch_export_data(ids) # Initialize queryset by calling base abstract function
+
+        self.queryset = self.queryset.values(
+            'id',
+            'name',
+            'container__name',
+            'container__barcode',
+            'coordinates',
+            'first_derived_sample',
+        )
+        samples = {s["id"]: s for s in self.queryset}
+
+        first_derived_sample_ids = self.queryset.values_list("first_derived_sample", flat=True)
+        derived_sample_values_queryset = (
+            DerivedSample.objects
+            .filter(id__in=first_derived_sample_ids)
+            .annotate(is_library=ExpressionWrapper(Q(library__isnull=False), output_field=BooleanField()))
+            .values(
+                'id',
+                'biosample__id',
+                'biosample__alias',
+                'project__name',
+            )
+        )
+        derived_samples = {ds["id"]: ds for ds in derived_sample_values_queryset}
+
+        biosample_ids = derived_sample_values_queryset.values_list('biosample__id', flat=True)
+        metadata_queryset = SampleMetadata.objects.filter(biosample_id__in=biosample_ids)
+        # For the renderer context
+        metadata_names = tuple(metadata_queryset.distinct().values_list('name', flat=True))
+        metadata_obj = metadata_queryset.values('biosample','name', 'value')
+
+        metadata_per_biosample = {}
+        for metadata in metadata_obj:
+            biosample = metadata['biosample']
+            if biosample in metadata_per_biosample.keys():
+                metadata_per_biosample[biosample].append(metadata)
+            else:
+                metadata_per_biosample[biosample] = [metadata] 
+
+        serialized_data = []
+        if not samples:
+            serialized_data.append({}) # Allow the returned csv file to be named instead of random name.
+        for sample in samples.values():
+            derived_sample = derived_samples[sample["first_derived_sample"]]
+            metadata = metadata_per_biosample[derived_sample["biosample__id"]]
+            
+            data = {
+                'alias': derived_sample["biosample__alias"],
+                'biosample_id': derived_sample["biosample__id"],     
+                'container_name': sample["container__name"],
+                'container_barcode': sample["container__barcode"],
+                'coordinates': sample["coordinates"], 
+                'project': derived_sample["project__name"],
+                **dict((item["name"], item["value"]) for item in metadata)
+            }
+            serialized_data.append(data)
+
+        return (metadata_names, serialized_data)
```

Comment:
metadata_names => data.keys() better to provide all fields and make sure we don't have mismatch between the fetch and the get_renderer_context.